platform/darwin/src/MGLOfflineStorage.mm

```diff
@@ -264,6 +264,81 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
     }
 }
 
+#pragma mark Offline merge methods
+
+- (void)addContentsOfFile:(NSString *)filePath withCompletionHandler:(MGLBatchedOfflinePackAdditionCompletionHandler)completion {
+    NSURL *fileURL = [NSURL fileURLWithPath:filePath];
+    
+    [self addContentsOfURL:fileURL withCompletionHandler:completion];
+
+}
+
+- (void)addContentsOfURL:(NSURL *)fileURL withCompletionHandler:(MGLBatchedOfflinePackAdditionCompletionHandler)completion {
+    
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    
+    if (!fileURL.isFileURL) {
+        [NSException raise:NSInvalidArgumentException format:@"%@ must be a valid file path", fileURL.absoluteString];
+    }
+    if (![fileManager isWritableFileAtPath:fileURL.path]) {
+        [NSException raise:NSInvalidArgumentException format:@"The file path: %@ must be writable", fileURL.absoluteString];
+    }
+    
+    __weak MGLOfflineStorage *weakSelf = self;
+    [self _addContentsOfFile:fileURL.path withCompletionHandler:^(NSArray<MGLOfflinePack *> * _Nullable packs, NSError * _Nullable error) {
+        if (packs) {
+            NSMutableDictionary *packsDictionary = [NSMutableDictionary dictionary];
+            NSMutableDictionary *packsIndex = [NSMutableDictionary dictionary];
+            NSInteger index = 0;
+            
+            MGLOfflineStorage *strongSelf = weakSelf;
+            for (MGLOfflinePack *pack in strongSelf.packs) {
+                [packsDictionary setObject:pack forKey:@(pack.mbglOfflineRegion->getID())];
+                [packsIndex setObject:@(index) forKey:@(pack.mbglOfflineRegion->getID())];
```

Comment:
We can avoid this redundant dictionary with the following approach:

```objc
// Map IDs to new packs.
NSMutableDictionary *packsByIdentifier = [NSMutableDictionary dictionary];
for (MGLOfflinePack *pack in packs) {
    [packsByIdentifier setObject:pack forKey:@(pack.mbglOfflineRegion->getID())];
}

id mutablePacks = [strongSelf mutableArrayValueForKey:@"packs"];
[strongSelf.packs enumerateObjectsUsingBlock:(^)(MGLOfflinePack *pack, NSUInteger idx, BOOL *stop) {
    MGLOfflinePack *newPack = packsByIdentifier[@(pack.mbglOfflineRegion->getID())];
    if (newPack) {
        [newPack invalidate];
        [mutablePacks replaceObjectAtIndex:idx withObject:pack];
        // By the end of this enumeration, packsByIdentifier only contains newly-added packs.
        packsByIdentifier[@(pack.mbglOfflineRegion->getID())] = nil;
    }
}];

// Add any remaining packs to the end of the array.
[mutablePacks addObjectsFromArray:packsByIdentifier.allValues];
```

To reduce the amount of work done via KVC, we could build up an NSIndexSet and an array of replacement packs, then pass both objects into `-replaceObjectsAtIndexes:withObjects:`. A further optimization would be to change `packsByIdentifier` into an `std::unordered_map`.

Reply:
> To reduce the amount of work done via KVC

Even though calling `-replaceObjectsAtIndexes:withObjects:` may be less elegant than the code above, it may be desirable because it prevents unnecessary thrashing on the part of key-value observers that would occur if each pack is replaced one by one.